validator_client/src/sync_committee_service.rs

```diff
@@ -0,0 +1,537 @@
+                            "error" => e
+                        )
+                    } else {
+                        trace!(
+                            log,
+                            "Spawned sync contribution tasks";
+                        )
+                    }
+
+                    // Do subscriptions for future slots/epochs.
+                    self.spawn_subscription_tasks();
+                } else {
+                    error!(log, "Failed to read slot clock");
+                    // If we can't read the slot clock, just wait another slot.
+                    sleep(slot_duration).await;
+                }
+            }
+        };
+
+        executor.spawn(interval_fut, "sync_committee_service");
+        Ok(())
+    }
+
+    async fn spawn_contribution_tasks(&self, slot_duration: Duration) -> Result<(), String> {
+        let log = self.context.log().clone();
+        let slot = self.slot_clock.now().ok_or("Failed to read slot clock")?;
+        let duration_to_next_slot = self
+            .slot_clock
+            .duration_to_next_slot()
+            .ok_or("Unable to determine duration to next slot")?;
+
+        // If a validator needs to publish a sync aggregate, they must do so at 2/3
+        // through the slot. This delay triggers at this time
+        let aggregate_production_instant = Instant::now()
+            + duration_to_next_slot
+                .checked_sub(slot_duration / 3)
+                .unwrap_or_else(|| Duration::from_secs(0));
+
+        let slot_duties = self
+            .duties_service
+            .sync_duties
+            .get_duties_for_slot::<E>(slot, &self.duties_service.spec)
+            .ok_or_else(|| format!("Error fetching duties for slot {}", slot))?;
+
+        if slot_duties.duties.is_empty() {
+            debug!(
+                log,
+                "No local validators in current sync committee";
+                "slot" => slot,
+            );
+            return Ok(());
+        }
+
+        // Fetch block root for `SyncCommitteeContribution`.
+        let block_root = self
+            .beacon_nodes
+            .first_success(RequireSynced::Yes, |beacon_node| async move {
+                beacon_node.get_beacon_blocks_root(BlockId::Head).await
+            })
+            .await
+            .map_err(|e| e.to_string())?
+            .ok_or_else(|| format!("No block root found for slot {}", slot))?
+            .data
+            .root;
+
+        // Spawn one task to publish all of the sync committee signatures.
+        let validator_duties = slot_duties.duties;
+        self.inner.context.executor.spawn(
+            self.clone()
+                .publish_sync_committee_signatures(slot, block_root, validator_duties)
+                .map(|_| ()),
+            "sync_committee_signature_publish",
+        );
+
+        let aggregators = slot_duties.aggregators;
+        self.inner.context.executor.spawn(
```

Comment:
Oh that's interesting, and quite subtle. I don't think I explicitly considered this selfish/altruistic tradeoff – I think it might have just come about as a side-effect of restructuring the attestation logic to support sync messages. The main change required was moving from a subnet-centric view of everything to _only_ caring about subnets for the aggregated messages (because the unaggregated messages need to be broadcast once, and then published to multiple subnets by the BN). That change split the two tasks into different `spawn` invocations, but it wouldn't be impossible to merge them together again.

The VC _should_ be able to publish the unaggregated messages in seconds 4-8 of the slot, so I'm tempted to leave this as-is for now, and reassess if we have issues?

Reply:
I agree with leaving this as-is for now. It's simpler, a rare edge-case (AFAICT), beneficial to the network and unlikely to be *that* bad for a validator.